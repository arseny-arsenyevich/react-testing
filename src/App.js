import React, { memo, useEffect, useRef, useState } from "react";
import { Routes, Route, Link } from "react-router-dom";
import MainPage from "./pages/MainPage";
import AboutPage from "./pages/AboutPage";
import ErrorPage from "./pages/ErrorPage";

function V() {
	const [state, setState] = useState(1);

	return (
		<>
			<button
				onClick={() => {
					setState(state + 1);
				}}
			>
				{state}
			</button>
			<h1>I am V</h1>
		</>
	);
}

function A() {
	const [state, setState] = useState(1);
	const ref = useRef();

	return (
		<div>
			I am div
			<button
				onClick={() => {
					setState(state + 1);
				}}
			>
				{state}
			</button>
			<p>
				I am p
				<V ref={ref} />
			</p>
		</div>
	);
}

function App() {
	const [data, setData] = useState(null)
	const [toggle, setToggle] = useState(false)
	const [value, setValue] = useState('')

	const onClick = () => setToggle(t => !t)

	useEffect(() => {
		setTimeout(() => {
			setData([])
		}, 100)
	}, [])


	return (
		<>
			{toggle === true && <div data-testid="toggle-div">toggle</div>}
			{data && <div>hello!</div>}
			<h1>работает</h1>
			<h1 data-testid="input-h1">{value}</h1>
			<button data-testid="toggle-btn" onClick={onClick}>click me</button>
			<input data-testid="input" type="text" placeholder="input" value={value} onChange={e => setValue(e.target.value)} />
			<Link to="/" data-testid="main-link">
				main
			</Link>
			<Link to="/about" data-testid="about-link">
				about
			</Link>
			<Routes>
				<Route path="/" Component={MainPage} />
				<Route path="/about" Component={AboutPage} />
				<Route path="/*" Component={ErrorPage} />
			</Routes>
		</>
	);
}

// console.log(<App />);

// function App() {
// 	const [data, setData] = useState(null)
// 	const [toggle, setToggle] = useState(false)
// 	const [value, setValue] = useState('')

// 	const onClick = () => setToggle(t => !t)

// 	useEffect(() => {
// 		setTimeout(() => {
// 			setData([])
// 		}, 100)
// 	}, [])

// 	return (
// 		<div className="App">
// 			{toggle === true && <div data-testid="toggle-div">toggle</div>}
// 			{data && <div>hello!</div>}
// 			<h1>работает</h1>
// 			<h1 data-testid="input-h1">{value}</h1>
// 			<button data-testid="toggle-btn" onClick={onClick}>click me</button>
// 			<input data-testid="input" type="text" placeholder="input" value={value} onChange={e => setValue(e.target.value)} />
// 		</div >
// 	);
// }

export default App;
