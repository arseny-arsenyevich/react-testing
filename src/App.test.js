import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import App from './App';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import createReduxStore from './store/store';

describe("test app", () => {
  test('renders learn react link', () => {
		render(
			<Provider store={createReduxStore()}>
				<MemoryRouter>
					<App />
				</MemoryRouter>
			</Provider>
		);
    const workELement = screen.getByText(/работает/i);
    const btn = screen.getByTestId("toggle-btn")
    const input = screen.getByPlaceholderText("input")
    expect(workELement).toBeInTheDocument();
    expect(btn).toBeInTheDocument();
    expect(input).toMatchSnapshot();

  });

  test('async render', async () => {
		render(
			<Provider store={createReduxStore()}>
				<MemoryRouter>
					<App />
				</MemoryRouter>
			</Provider>
		);
    const helloDiv = await screen.findByText(/hello!/i);
    expect(helloDiv).toBeInTheDocument()
  });

  test('click event', async () => {
		render(
			<Provider store={createReduxStore()}>
				<MemoryRouter>
					<App />
				</MemoryRouter>
			</Provider>
		);
    const btn = screen.getByTestId("toggle-btn");
    let toggleDiv = screen.queryByTestId("toggle-div");
    expect(toggleDiv).toBeNull()
    fireEvent.click(btn)
    toggleDiv = screen.queryByTestId("toggle-div");
    expect(toggleDiv).toBeInTheDocument()
    fireEvent.click(btn)
    toggleDiv = screen.queryByTestId("toggle-div");
    expect(toggleDiv).toBeNull()
  });

  test('click event 2', async () => {
		render(
			<Provider store={createReduxStore()}>
				<MemoryRouter>
					<App />
				</MemoryRouter>
			</Provider>
		);
    const input = screen.getByTestId("input");
    let inputH = screen.queryByTestId("input-h1");
    fireEvent.input(input, {
      target: { value: '123' }
    })

    // ну так се проверочка
    expect(inputH).toContainHTML("2")
  });
})
