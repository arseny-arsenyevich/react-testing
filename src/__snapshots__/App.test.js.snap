// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test app renders learn react link 1`] = `
<input
  data-testid="input"
  placeholder="input"
  type="text"
  value=""
/>
`;
